junit-platform-commons/src/main/java/org/junit/platform/commons/support/conversion/TypeDescriptor.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2015-2025 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.platform.commons.support.conversion;
+
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Parameter;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Optional;
+
+import org.apiguardian.api.API;
+import org.jspecify.annotations.Nullable;
+import org.junit.platform.commons.util.Preconditions;
+import org.junit.platform.commons.util.ReflectionUtils;
+
+/**
+ *
+ *
+ * @since 6.0
+ */
+@API(status = EXPERIMENTAL, since = "6.0")
+public final class TypeDescriptor {
+
+	/**
+	 * {@code TypeDescriptor} returned when no value is available.
+	 */
+	public static final TypeDescriptor NONE = new TypeDescriptor(null);
+
+	private final @Nullable Class<?> type;
+
+	public static TypeDescriptor forClass(Class<?> clazz) {
+		Preconditions.notNull(clazz, () -> "clazz must not be null");
+		return new TypeDescriptor(clazz);
+	}
+
+	public static TypeDescriptor forInstance(@Nullable Object instance) {
+		return instance != null ? forClass(instance.getClass()) : NONE;
+	}
+
+	public static TypeDescriptor forField(Field field) {
+		Preconditions.notNull(field, "field must not be null");
+		return forClass(field.getType());
+	}
+
+	public static TypeDescriptor forParameter(Parameter parameter) {
+		Preconditions.notNull(parameter, "parameter must not be null");
+		return forClass(parameter.getType());
+	}
+
+	private TypeDescriptor(@Nullable Class<?> type) {
+		this.type = type;
+	}
+
+	public Class<?> getType() {
```

Comment:
The discussion in #5004 made me think that we should consider supporting `Type` instead of just `Class<?>`, i.e. making this class more like Guava's [`TypeToken`](https://guava.dev/releases/21.0/api/docs/com/google/common/reflect/TypeToken.html) or Spring's [`ParameterizedTypeReference`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/core/ParameterizedTypeReference.html).

Thoughts?

Reply:
> Example: `"1,2,3"` as `source` and `List<Integer>` vs. `List<Long>` as `targetType`

I got a similar use case covered in scordio/junit-converters#4 (see the corresponding test case [here](https://github.com/scordio/junit-converters/blob/c8268bf9f5045fe8722ff4dd35857de8002e6236/src/test/java/io/github/scordio/junit/converters/tests/SpringConversionIntegrationTests.java#L49)).

However, that's at the level of `ArgumentConverter` where `Field` or `Parameter` is directly available and can be fed into the Spring classes to perform the conversion.

I'll keep this use case in mind and see how the Spring integration can work with the changes from this PR, especially to determine whether the current `TypeDescriptor` APIs are sufficient for this purpose.